### PR TITLE
[SPARK-34595][SQL] DPP support RLIKE

### PR DESCRIPTION
**Pull request by chaojun-zhang**
### What changes were proposed in this pull request?
This pr make DPP support RLIKE expression:

```sql
SELECT date_id, product_id FROM fact_sk f
JOIN dim_store s
ON f.store_id = s.store_id WHERE s.country RLIKE  '[DE|US]'
``` 
 ### Why are the changes needed?
Improve query performance.
 
 ### Does this PR introduce _any_ user-facing change?
 No.
 
 ### How was this patch tested?
 Unit test.

### CODE REVIEW

**Comment by wangyum** [comment]
ok to test.

**Comment by HyukjinKwon** [comment]
ok to test

File: sql/core/src/test/scala/org/apache/spark/sql/DynamicPartitionPruningSuite.scala [hunk 1]
```diff
@@ -1403,6 +1403,50 @@ abstract class DynamicPartitionPruningSuiteBase
...
+        """.stripMargin)
+
+      checkPartitionPruningPredicate(df, false, true)
+
+      checkAnswer(df,
+        Row(1030, 2) ::
+          Row(1040, 2) ::
+          Row(1050, 2) ::
+          Row(1060, 2) :: Nil
+      )
```
Comment (maropu) [inline sql/core/src/test/scala/org/apache/spark/sql/DynamicPartitionPruningSuite.scala:1423]: We already have this LIKE-case test in the existing code? If it exists, I think we don't need this part.

Comment (chaojun-zhang) [reply sql/core/src/test/scala/org/apache/spark/sql/DynamicPartitionPruningSuite.scala]: Ok, I will remove this part

Comment (chaojun-zhang) [reply sql/core/src/test/scala/org/apache/spark/sql/DynamicPartitionPruningSuite.scala]: Fix

File: sql/core/src/test/scala/org/apache/spark/sql/DynamicPartitionPruningSuite.scala [hunk 2]
```diff
@@ -1403,6 +1403,32 @@ abstract class DynamicPartitionPruningSuiteBase
...
     }
   }
+
+  test("SPARK-34436: DPP support RLike expression") {
+    withSQLConf(SQLConf.DYNAMIC_PARTITION_PRUNING_ENABLED.key -> "true") {
+      val df = sql(
+        """
+          |SELECT date_id, product_id FROM fact_sk f
+          |JOIN dim_store s
+          |ON f.store_id = s.store_id WHERE s.country RLIKE  '[DE|US]'
```
Comment (c21) [inline sql/core/src/test/scala/org/apache/spark/sql/DynamicPartitionPruningSuite.scala:1413]: nit: extra unnecessary spaces between `RLIKE` and `'[DE...`

File: sql/core/src/test/scala/org/apache/spark/sql/DynamicPartitionPruningSuite.scala [hunk 3]
```diff
@@ -1403,6 +1403,32 @@ abstract class DynamicPartitionPruningSuiteBase
...
+        Row(1030, 2) ::
+          Row(1040, 2) ::
+          Row(1050, 2) ::
+          Row(1060, 2) ::
+          Row(1070, 2) ::
+          Row(1080, 3) ::
+          Row(1090, 3) ::
+          Row(1100, 3) ::
+          Row(1110, 3) ::
+          Row(1120, 4) :: Nil
```
Comment (c21) [inline sql/core/src/test/scala/org/apache/spark/sql/DynamicPartitionPruningSuite.scala:1428]: nit: should be same indentation as other test cases in this file.

**Review by maropu — APPROVED** [review]

**Review by c21 — COMMENTED** [review]
LGTM with minor style comments.

File: sql/core/src/test/scala/org/apache/spark/sql/DynamicPartitionPruningSuite.scala (hunk 2 above)
Comment (chaojun-zhang) [reply sql/core/src/test/scala/org/apache/spark/sql/DynamicPartitionPruningSuite.scala]: Updated

File: sql/core/src/test/scala/org/apache/spark/sql/DynamicPartitionPruningSuite.scala (hunk 3 above)
Comment (chaojun-zhang) [reply sql/core/src/test/scala/org/apache/spark/sql/DynamicPartitionPruningSuite.scala]: Updated

**Comment by HyukjinKwon** [comment]
Looks fine to me too. It would be great if @maryannxue could have a chance to take a quick look.

**Review by c21 — APPROVED** [review]
LGTM

**Review by HyukjinKwon — APPROVED** [review]

**Review by viirya — APPROVED** [review]

**Comment by maropu** [comment]
Thanks! Merged to master.

**Comment by maryannxue** [comment]
Thank for pinging me, @HyukjinKwon ! Late LGTM here.

**Pull request closed without merging**